### PR TITLE
Tracking: Transport, Relay, and Discovery Enhancements 

### LINKED ISSUE #691: Tracking: Transport, Relay, and Discovery Enhancements
### Description

This tracking issue covers multiple enhancements across the `libp2p` stack to improve how peers connect, route through relays, and discover each other. The goal is to make the networking layer more resilient, especially in NATed or firewalled environments, while improving correctness and modularity.

#### Scope includes:

* Secure relay reservation and maintenance
* Voucher and signature validation
* Smarter relay selection algorithms
* Proper handling of relayed `/p2p-circuit` multiaddrs
* DHT-based peer discovery integration
* Accurate peer metadata including relay multiaddrs

Each task listed below will be tracked via sub-issues and pull requests. Contributors can reference this issue in their PRs using `Closes #<this-issue>` where appropriate.

### Motivation

The primary motivation behind this work is to evolve `py-libp2p` into a **production-grade library**, capable of supporting real-world deployments in decentralized applications, research networks, and production systems.

This direction has been discussed and aligned with @paul (and others), and reflects the need for:

* **Stability and correctness** under real-world network conditions
* **Parity with go-libp2p / js-libp2p** in key features
* **Support for multi-hop, relayed, and NATed connectivity**
* **A modular, extensible architecture** for experimentation and new protocols

To reach this production-ready state, several key features must be implemented, hardened, and tested:

1. **Reliable relay support** – with reservation refresh, multiaddr correctness, and flexible selection logic
2. **Secure resource validation** – with voucher and signature verification to guard access and allocation
3. **Dynamic peer discovery** – using the DHT to locate peers in disconnected or semi-connected topologies
4. **Proper address handling** – including encoding/decoding of relayed multiaddrs, and updating peer metadata

These enhancements not only improve developer and user experience but also unblock downstream use cases such as:

* Building permissionless P2P networks in constrained environments
* Research and prototyping of new p2p protocols in Python

### 📋 **TODOs**

* [ ] **Implement voucher and signature verification in `resources.py`**
  Validate incoming relay vouchers and signatures to ensure proper authorization and prevent misuse.

* [ ] **Implement initial relay selection logic in `transport.py`**
  Use basic selection (e.g., first-available or round-robin) to choose a relay when dialing.

* [ ] **Implement DHT-based peer discovery**
  Use the libp2p DHT to dynamically locate and connect to peers across the network.

* [ ] **Implement more sophisticated relay selection**
  Introduce scoring, latency-based metrics, or retry strategies for better relay routing.

* [ ] **Implement reservation storage and refresh mechanism**
  Track active relay reservations and refresh them before expiry to support long-lived relayed connections.

* [ ] **Add relay multiaddrs to peerstore**
  Store `/p2p-circuit/...` addresses in the peer metadata to support reconnects and discovery.

* [ ] **Implement proper multiaddr parsing and handling for relayed connections**
  Ensure `/p2p-circuit/p2p/...` paths are correctly parsed, validated, and used during dialing.

### Open questions

_No response_

### Are you planning to do it yourself in a pull request ?

Maybe

### PR DESCRIPTION
**Pull request by codemaestro64**
## What was wrong?

Issue #691

The libp2p stack lacked robust relay support, secure resource validation, and dynamic peer discovery, impacting reliability in NATed or firewalled environments.

## Description

This PR enhances the py-libp2p networking layer with improvements to relay functionality and security, advancing the library toward production-grade reliability for decentralized applications.

## TODOs
- [x] Implement voucher and signature verification in `resources.py`
- [x] Implement initial relay selection logic in `transport.py`
- [x] Implement DHT-based peer discovery
- [x] Implement more sophisticated relay selection
- [x] Implement reservation storage and refresh mechanism
- [x] Add relay multiaddrs to peerstore
- [x] Implement proper multiaddr parsing and handling for relayed connections
- [x] Implement `run()` method in `CircuitV2Listener`

closes #691

### CODE REVIEW

**Comment by seetadev** [comment]
@codemaestro64 : Thank you so much for opening the PR for completing all the required enhancements and todos in circuit relay. Appreciate your efforts.

Updated the branch and re-run CI/CD pipeline. You will get immediate results on CI/CD test cases.

**Comment by seetadev** [comment]
@codemaestro64 : Appreciate your efforts.

Wish to invite @acul71 , @lla-dane, @Winter-Soren for sharing feedback points on the PR.

**Comment by Winter-Soren** [comment]
Hey @codemaestro64,
I have reviewed your PR.
delivers production‑grade rigor: secure reservation validation, resilient relay discovery/selection, durable reservation management, and correct circuit multiaddr behavior. The protocol/service lifecycle is implemented cleanly with timeouts and explicit error signaling. Tests cover the core and many edge conditions. This is a well‑structured, merge‑ready upgrade that improves reliability and parity with other libp2p stacks.

CC'ing @seetadev @acul71

File: libp2p/relay/circuit_v2/transport.py [hunk 1]
```diff
@@ -154,6 +168,36 @@ async def dial_peer_info(
             If the connection cannot be established
 
         """
+        # Prefer stored /p2p-circuit addrs from peerstore
+        # Try first to read addresses from peerstore
+        peer_store = self.host.get_peerstore()
+        stored_addrs = peer_store.addrs(peer_info.peer_id)
```
Comment (pacrob) [inline libp2p/relay/circuit_v2/transport.py]: `peer_store.addrs` will raise a `PeerStoreError` if the peer we're dialing isn't already in the peerstore, making dialing new peers impossible. Maybe wrap in a `try/except` or similar?

**Comment by seetadev** [comment]
@codemaestro64 : Great work. Appreciate the progress.

Please reply to @pacrob so that he can review and mark the feedback as resolved.

There is one CI/CD failure, which can be ignored for now (not related to the PR):

=========================== short test summary info ============================
FAILED tests/core/kad_dht/test_kad_dht.py::test_put_and_get_value - Assertion...
============ 1 failed, 1551 passed, 5 warnings in 553.16s (0:09:13) ============
py311-core: exit 1 (554.53 seconds) /home/runner/work/py-libp2p/py-libp2p> pytest -n auto --timeout=1200 tests/core pid=2376
  py311-core: FAIL code 1 (581.97=setup[27.44]+cmd[554.53] seconds)
  evaluation failed :( (582.54 seconds)
Error: Process completed with exit code 1.

**Comment by seetadev** [comment]
@codemaestro64 : Wonderful efforts indeed. Appreciate your contribution.

Wish if you could mark the todos at https://github.com/libp2p/py-libp2p/issues/691 completed in your PR. We are reviewing the PR in details, once again. It is indeed heading towards final review + merge. Great progress.

On the same note, wish if you could open up a discussion page sharing more details on the implementation steps and test suite.

CCing @pacrob, @Winter-Soren, who are reviewing this PR. Looking forward to their pointers and feedback.

File: pyproject.toml [hunk 2]
```diff
@@ -283,6 +285,9 @@ force-to-top = ["pytest"]
 # Using Ruff's Black-compatible formatter.
 # Options like quote-style = "double" or indent-style = "space" can be set here if needed.
 
+[tool.ruff.per-file-ignores]
+"tests/*.py" = ["F821"]  # undefined name (e.g., pytest fixtures like mocker)
+
```
Comment (pacrob) [inline pyproject.toml]: I'm not getting an error with this. Is it necessary?

File: libp2p/relay/circuit_v2/transport.py [hunk 3]
```diff
@@ -436,10 +841,92 @@ def __init__(
         self.multiaddrs: list[
             multiaddr.Multiaddr
         ] = []  # Store multiaddrs as Multiaddr objects
+        self.handler_function = handler_function
+
+    async def handle_incoming_connection(
+        self,
+        stream: INetStream,
+        remote_peer_id: ID,
```
Comment (pacrob) [inline libp2p/relay/circuit_v2/transport.py]: This arg is never used in the function.

Comment (codemaestro64) [reply libp2p/relay/circuit_v2/transport.py]: My bad. It's been removed

File: pyproject.toml (hunk 2 above)
Comment (codemaestro64) [reply pyproject.toml]: No longer needed. Removed

File: libp2p/relay/circuit_v2/protocol.py
Comment (pacrob) [inline libp2p/relay/circuit_v2/protocol.py:602]: Should these be updated? It looks like we are checking against these values in `RelayResourceManager->verify_reservation`, but they are always empty strings here.

**Comment by seetadev** [comment]
@codemaestro64 : Wish if you could address the improvements shared by @pacrob. Appreciate your contribution.

Wish if you also could discuss with @Winter-Soren and @sumanjeet0012 and get their pointers while addressing important feedback points by @pacrob.

File: libp2p/relay/circuit_v2/protocol.py
Comment (codemaestro64) [reply libp2p/relay/circuit_v2/protocol.py:602]: @pacrob Vouchers haven't been implemented yet. @seetadev I'd be happy to pick that up as soon as you give the word. But I suggest this PR be reviewed and merged as this and the Voucher implementation are of two scopes

Comment (seetadev) [reply libp2p/relay/circuit_v2/protocol.py:602]: @codemaestro64 : Thank you for sharing your feedback on voucher implementation.

Vouchers have been implemented. In https://github.com/libp2p/py-libp2p/pull/698, @guha-rahul fixed voucher verification and data transfer limits in Circuit Relay v2. We are just merging that PR in a day or so.

Wish to invite @sukhman-sukh, @Winter-Soren, @acul71 and @guha-rahul to also do a peer review of the features added in this PR. 

Wish to ensure that we have all the requisite features in circuit relay as discussed in maintainer's meeting.

**Review by pacrob — APPROVED** [review]
lgtm!

**Comment by seetadev** [comment]
@pacrob : Hi Paul. Thank you so much for your feedback. Appreciate it.

This PR is indeed ready for final review + merged.

@codemaestro64: Wish to recommend you to open a discussion page and share a note for other NAT traversal developers on your implementation details in this PR, test suite. We will like you to follow up this PR with an example demonstrating real use-case of circuit relay and other NAT traversal components in a project.
